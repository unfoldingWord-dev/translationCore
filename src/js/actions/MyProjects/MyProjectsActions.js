import consts from '../ActionTypes';
// helpers
import * as myProjectsHelpers from '../../helpers/myProjectsHelpers';

/**
 * @description With the list of project directories, generates an array of project detail objects
 */
export function getMyProjects() {
  return ((dispatch, getState) => {
    myProjectsHelpers.migrateResourcesFolder();
    const state = getState();
    const { projectDetailsReducer: { projectSaveLocation } } = state;
    let projects = myProjectsHelpers.getProjectsFromFS(projectSaveLocation);
    dispatch({
      type: consts.GET_MY_PROJECTS,
      projects: projects
    });
  });
}
