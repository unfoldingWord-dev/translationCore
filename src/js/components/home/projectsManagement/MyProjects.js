import React from 'react';
import PropTypes from 'prop-types';
// components
import ProjectCard from './ProjectCard';

let MyProjects = (props) => {
  const {myProjects} = props;
  const projects = myProjects.map( (projectDetails, index) =>
    <ProjectCard key={index} projectDetails={projectDetails} actions={props.actions} />
  );

  return (
    <div style={{ height: '100%' }}>
      Projects
      <div style={{ height: '100%', overflowY: 'auto', paddingRight: '10px' }}>
        {projects}
      </div>
    </div>
  );
}

MyProjects.propTypes = {
  myProjects: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired
};

export default MyProjects;
