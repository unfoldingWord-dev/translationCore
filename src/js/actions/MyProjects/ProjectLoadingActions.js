import consts from '../ActionTypes';
import path from 'path-extra';
// actions
import * as ProjectMigrationActions from '../Import/ProjectMigrationActions';
import * as ProjectValidationActions from '../Import/ProjectValidationActions';
import * as ToolsMetadataActions from '../ToolsMetadataActions';
import * as BodyUIActions from '../BodyUIActions';
import * as RecentProjectsActions from '../RecentProjectsActions';
import * as AlertModalActions from '../AlertModalActions';
import * as ProjectDetailsActions from '../ProjectDetailsActions';
import * as ProjectLoadingActions from '../MyProjects/ProjectLoadingActions';
import * as ProjectImportStepperActions from '../ProjectImportStepperActions';
//helpers
import * as manifestHelpers from '../../helpers/manifestHelpers';
// constants
const PROJECTS_PATH = path.join(path.homedir(), 'translationCore', 'projects');

/**
 * @description action that Migrates, Validates and Loads the Project
 * This may seem redundant to run migrations and validations each time
 * But the helpers called from each action test to only run when needed
 * @param {String} selectedProjectFilename
 */
export const migrateValidateLoadProject = (selectedProjectFilename) => {
  return(async (dispatch) => {
    try {
      dispatch(AlertModalActions.openAlertDialog('Loading your project data', true));
      setTimeout(async () => {
        let projectPath = path.join(PROJECTS_PATH, selectedProjectFilename);
        ProjectMigrationActions.migrate(projectPath);
        dispatch(AlertModalActions.closeAlertDialog());
        await dispatch(ProjectValidationActions.validate(projectPath));
        dispatch(displayTools());
      }, 200);
    } catch (error) {
      if (error.type !== 'div') console.warn(error);
      // clear last project must be called before any other action.
      // to avoid triggering autosaving.
      dispatch(clearLastProject());
      dispatch(AlertModalActions.openAlertDialog(error));
      dispatch(ProjectImportStepperActions.cancelProjectValidationStepper());
      dispatch({ type: "LOADED_ONLINE_FAILED" });
    }
  });
};

/**
 * @description - Opening the tools screen upon making sure the project is
 * not a titus or in the user is in developer
 */
export function displayTools() {
  return ((dispatch, getState) => {
    const { currentSettings } = getState().settingsReducer;
    const { manifest } = getState().projectDetailsReducer;
    if (manifestHelpers.checkIfValidBetaProject(manifest) || currentSettings.developerMode) {
      dispatch(ToolsMetadataActions.getToolsMetadatas());
      // Go to toolsCards page
      dispatch(BodyUIActions.goToStep(3));
    } else {
      dispatch(AlertModalActions.openAlertDialog('This version of translationCore only supports Titus projects.'));
      dispatch(RecentProjectsActions.getProjectsFromFolder());
    }
  });
}

/**
 * @description - Wrapper to clear everything in the store that could
 * prevent a new project from loading
 */
export function clearLastProject() {
  return ((dispatch) => {
    dispatch(BodyUIActions.toggleHomeView(true));
    dispatch(ProjectDetailsActions.resetProjectDetail());
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_DATA });
    dispatch({ type: consts.CLEAR_PREVIOUS_GROUPS_INDEX });
    dispatch({ type: consts.CLEAR_CONTEXT_ID });
    dispatch({ type: consts.CLEAR_CURRENT_TOOL_DATA });
    dispatch({ type: consts.CLEAR_RESOURCES_REDUCER });
    dispatch({
      type: consts.SET_CURRENT_TOOL_TITLE,
      currentToolTitle: ""
    });
    /** After clearing the local project the label also needs to be updated in the stepper */
    dispatch(BodyUIActions.resetStepLabels(1));
  });
}

/**
 * @description loads and set the projects details into the projectDetailsReducer.
 * @param {string} projectPath - path location in the filesystem for the project.
 * @param {object} manifest - project manifest.
 */
export function loadProjectDetails(projectPath, manifest) {
  return ((dispatch) => {
    dispatch(ProjectDetailsActions.setSaveLocation(projectPath));
    dispatch(ProjectDetailsActions.setProjectManifest(manifest));
  });
}
