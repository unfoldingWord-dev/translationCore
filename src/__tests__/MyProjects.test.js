/* eslint-env jest */
import React from 'react';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { shallow } from 'enzyme';
import MyProjects from '../js/components/home/projectsManagement/MyProjects';

// Tests for MyProjects React Component
describe('MyProjects component renders correctly', () => {
  test('MyProjects component render should match snapshot', () => {
    const myProjects = [
      {
        projectName: 'en_1co_ulb',
        projectSaveLocation: '<TEMP_DIR>/en_1co_ulb',
        accessTimeAgo: '5 days ago',
        bookAbbr: '1co',
        bookName: '1 Corinthians',
        target_language: {
          id: 'en',
          name: 'English',
        },
        isSelected: false,
      },
      {
        projectName: 'hi_tit',
        projectSaveLocation: '<TEMP_DIR>/hi_tit',
        accessTimeAgo: '6 days ago',
        bookAbbr: 'tit',
        bookName: 'Titus',
        target_language: {
          id: 'hi',
          name: 'Hindi',
        },
        isSelected: false,
      },
    ];
    const userdata = { username: 'manny-test' };
    const renderedValue = shallow(
      <MuiThemeProvider>
        <MyProjects
          translate={key => key}
          myProjects={myProjects}
          user={userdata}
          onSelect={() => jest.fn()} />
      </MuiThemeProvider>,
    );

    expect(renderedValue).toMatchSnapshot();
  });
});

describe('MyProjects shows instruction if no projects', () => {
  test('MyProjects component render should match snapshot', () => {
    const myProjects = [];
    const userdata = { username: 'manny-test' };
    const blank = shallow(
      <MuiThemeProvider>
        <MyProjects
          translate={key => key}
          myProjects={myProjects}
          user={userdata}
          onSelect={() => jest.fn()} />
      </MuiThemeProvider>,
    );

    expect(blank).toMatchSnapshot();
  });
});
