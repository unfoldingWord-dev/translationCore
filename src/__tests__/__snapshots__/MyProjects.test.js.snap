// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyProjects component renders correctly MyProjects component render should match snapshot 1`] = `ShallowWrapper {}`;

exports[`MyProjects shows instruction if no projects MyProjects component render should match snapshot 1`] = `ShallowWrapper {}`;
